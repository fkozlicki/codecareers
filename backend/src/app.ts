import bodyParser from 'body-parser';
import cors from 'cors';
import dotenv from 'dotenv';
import express, { Request, Response } from 'express';
import { corsOptions } from './config/corsOptions';
import { serveFile } from './lib/s3';
import { authorizeCv } from './middleware/authorization';
import { requireSession, verifySession } from './middleware/session';
import { applicationsRouter } from './routes/application.routes';
import { authRouter } from './routes/auth.routes';
import { chatsRouter } from './routes/chat.routes';
import { companiesRouter } from './routes/company.routes';
import { jobOffersRouter } from './routes/jobOffer.routes';
import { recruitmentsRouter } from './routes/recruitment.routes';
import { skillsRouter } from './routes/skill.routes';
import { technologiesRouter } from './routes/technology.routes';
import { usersRouter } from './routes/user.routes';

dotenv.config({ path: `.env.${process.env.NODE_ENV ?? 'development'}` });

const app = express();

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));
app.use(cors(corsOptions));

app.use(verifySession);

app.get('/', (req: Request, res: Response) => {
	res.json({ message: 'Welcome to CodeCareers API' });
});

app.use('/', authRouter);
app.use('/companies', companiesRouter);
app.use('/job-offers', jobOffersRouter);
app.use('/technologies', technologiesRouter);
app.use('/skills', skillsRouter);
app.use('/applications', applicationsRouter);
app.use('/users', usersRouter);
app.use('/recruitments', recruitmentsRouter);
app.use('/chats', chatsRouter);
app.use('/avatars/:filename', serveFile('avatars'));
app.use('/cv/:filename', requireSession, authorizeCv, serveFile('cvs'));

export default app;
