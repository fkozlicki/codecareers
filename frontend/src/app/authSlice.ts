import { User, session } from '@/app/services/auth';
import { PayloadAction, createSlice } from '@reduxjs/toolkit';

interface AuthState {
	user: User | null;
	status: 'loading' | 'authenticated' | 'unauthenticated';
}

const initialState: AuthState = {
	user: null,
	status: 'loading',
};

const { reducer, actions } = createSlice({
	name: 'auth',
	initialState,
	reducers: {
		logout: () => initialState,
		setUser: (state, action: PayloadAction<User>) => {
			state.user = action.payload;
		},
	},
	extraReducers: (builder) => {
		builder
			.addMatcher(session.matchPending, (state) => {
				state.status = 'loading';
			})
			.addMatcher(session.matchFulfilled, (state, action) => {
				if (action.payload.user) {
					state.status = 'authenticated';
					state.user = action.payload.user;
				} else {
					state.status = 'unauthenticated';
				}
			});
	},
});

export const { logout, setUser } = actions;

export default reducer;
