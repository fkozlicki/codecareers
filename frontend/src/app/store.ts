import auth from '@/app/authSlice';
import { api } from '@/app/services/api';
import { combineReducers, configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';

const rootReducer = combineReducers({
	[api.reducerPath]: api.reducer,
	auth,
});

export const setupStore = (preloadedState?: Partial<RootState>) => {
	return configureStore({
		reducer: rootReducer,
		middleware: (getDefaultMiddleware) =>
			getDefaultMiddleware().concat(api.middleware),
		preloadedState,
	});
};

export const store = setupStore();

setupListeners(store.dispatch);

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof rootReducer>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
export type AppStore = ReturnType<typeof setupStore>;
