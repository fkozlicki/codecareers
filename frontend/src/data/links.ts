import {
	Building2Icon,
	FilePenIcon,
	LogOutIcon,
	SettingsIcon,
	UsersIcon,
} from 'lucide-react';

export const navigationLinks = [
	{
		label: 'Applications',
		Icon: FilePenIcon,
		href: '/my-applications',
	},
	{
		label: 'Recruitments',
		Icon: UsersIcon,
		href: '/my-recruitments',
	},
	{
		label: 'Companies',
		Icon: Building2Icon,
		href: '/my-companies',
	},
	{
		label: 'Settings',
		Icon: SettingsIcon,
		href: '/settings',
	},
	{
		label: 'Logout',
		Icon: LogOutIcon,
		href: `${import.meta.env.VITE_API_URI}/logout`,
	},
];
