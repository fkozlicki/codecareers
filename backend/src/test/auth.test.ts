import supertest from 'supertest';
import app from '../app.js';
import { cleanupDB, initializeDB, password } from './data.js';

beforeEach(async () => {
	await initializeDB();
});

afterEach(async () => {
	await cleanupDB();
});

describe('Auth Service', () => {
	describe('[POST] /signup', () => {
		it('Create new user, when credentials are valid and no user with the same email exists', async () => {
			const { statusCode, body } = await supertest(app).post('/signup').send({
				firstName: 'Anna',
				lastName: 'Marie',
				email: 'annamarie@gmail.com',
				password: 'annamarie123',
			});

			expect(statusCode).toBe(201);
			expect(body).toEqual({ message: 'Signed up successfully' });
		});

		it('When user with same email exists, throw a 409 (duplicate) error', async () => {
			const { statusCode, body } = await supertest(app).post('/signup').send({
				firstName: 'Jon',
				lastName: 'Snow',
				email: 'jon.snow@gmail.com',
				password: 'jonsnow123',
			});

			expect(statusCode).toBe(409);
			expect(body).toEqual({ message: 'User with this email already exists' });
		});

		it('Throw 400 when data is invalid', async () => {
			const invalidCredentials = [
				// missing lastName
				{
					firstName: 'Lilly',
					email: 'lilly@gmail.com',
					password: 'lilly123',
				},
				// missing firstName
				{
					lastName: 'Pietrova',
					email: 'pietrova@gmail.com',
					password: 'pietrova123',
				},
				// missing email
				{
					firstName: 'Alan',
					lastName: 'Walker',
					password: 'alanwalker123',
				},
				// missing password
				{
					firstName: 'Jake',
					lastName: 'Paul',
					email: 'jake.paul@gmail.com',
				},
			];

			for (let credentials of invalidCredentials) {
				const { statusCode, body } = await supertest(app)
					.post('/signup')
					.send(credentials);

				expect(statusCode).toBe(400);
				expect(body.message).toEqual('Invalid data');
			}
		});
	});

	describe('[POST] /login/credentials', () => {
		it('Sign in, when credentials are valid', async () => {
			const { statusCode, body } = await supertest(app)
				.post('/login/credentials')
				.send({
					email: 'jon.snow@gmail.com',
					password,
				});

			expect(statusCode).toBe(200);
			expect(body.user).toMatchSnapshot({
				id: expect.any(String),
			});
		});

		it('Throw 400, when data is invalid', async () => {
			const invalidCredentials = [
				// missing email
				{
					password: 'alanwalker123',
				},
				// missing password
				{
					email: 'jake.paul@gmail.com',
				},
			];

			for (let credentials of invalidCredentials) {
				const { statusCode, body } = await supertest(app)
					.post('/login/credentials')
					.send(credentials);

				expect(statusCode).toBe(400);
				expect(body.message).toEqual('Invalid data');
			}
		});
	});
});
