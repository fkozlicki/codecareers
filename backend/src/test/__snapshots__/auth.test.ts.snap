// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Service [POST] /login/credentials Sign in, when credentials are valid 1`] = `
{
  "avatar": null,
  "email": "jon.snow@gmail.com",
  "firstName": "Jon",
  "githubId": null,
  "id": Any<String>,
  "lastName": "Snow",
  "username": null,
}
`;
